Mu2eG4/inc/Mu2eG4PerThreadStorage.hh

```diff
@@ -78,10 +79,14 @@ namespace mu2e {
 
     bool runTerminated = false;
 
+    // this comes in from subRuns, via the physical volume info
+    unsigned simStage = -1u;
+
     // event-level data members
     art::Event* artEvent = nullptr;
-    SimParticleHelper simParticleHelper;
-    SimParticlePrimaryHelper simParticlePrimaryHelper;
+    // delay initialization of the helpers using std::optional
+    std::optional<SimParticleHelper> simParticleHelper;
+    std::optional<SimParticlePrimaryHelper> simParticlePrimaryHelper;
 
```

Comment:
Do we know that std::optional is thread safe?  It seems like it should be since it appears to work in thread local storage.

Reply:
I think I saw it documented as thread-safe somewhere when I researched the construct, but can't find the reference now.

Reply:
I think I saw it documented as thread-safe somewhere when I researched the construct, but can't find the reference now.

Reply:
Actually, std::optional thread safety should not matter here.  The top level structure is only accessed by one thread, the "per-thread" storage is in its name.    Looking at a draft of the C++ standard, std::optional is about as thread-safe as unique_ptr.  Or an int.  None of them can be simultaneously written by different threads.

    Any instance of optional<T> at any given time either contains a value
    or does not contain a value. When an instance of optional<T> contains
    a value, it means that an object of type T, referred to as the
    optional object’s contained value, is allocated within the storage of
    the optional object. Implementations are not permitted to use
    additional storage, such as dynamic memory, to allocate its contained
    value. The contained value shall be allocated in a region of the
    optional<T> storage suitably aligned for the type T. When an object of
    type optional<T> is contextually converted to bool, the conversion
    returns true if the object contains a value; otherwise the conversion
    returns false.

Reply:
Thanks.   This is the first use that I have seen of std::optional.   Here is how I am thinking of the difference between the two.   We could have used std::unique_ptr and had exactly the same behaviour - we only care about delayed instantiation and an automated call to the d'tor at the right time.   The difference is that std::optional clearly says that we never plan to transfer ownership of the object; on the other hand std::unique_ptr allows that.  Is that how you see it?

If so we should plan to transition this use of std::unique_ptr to std::optional as targets of opportunity.